### ISSUE TITLE
`NaN` loss value occurs in training process

### ISSUE THREAD

**Issue by hermosaaurora** [Wanggcong/SparseNeRF#6]
hi, I try to use the training code. But I found after about 2k~5k iterations, it will run into a `nan` loss. And then I found some tricks in your repo.

1. set the `config.check_grad_for_nans` to be `True`
2. you implemented some safe function versions for `jnp.xxx` functions such as`safe_sin` and `safe_cos`

I'm not digging much into the phenomenon now (maybe later I gotta do that), but is the nan loss common to appear in training process for your algorithms? Can you give some advice on why it will happen?

upd: I guess it is caused by something like "divided by zero", because I found the config.txt has `Config.grad_max_val` and `Config.grad_max_norm` setting and the code will do some clipping to the grad. So there must be sometime when the loss gets a `nan` suddenly, and the clipping just fails.

**Comment by hermosaaurora** [comment]
Besides, I'm using my own dataset, not the example data you provided.

**Comment by Wanggcong** [comment]
Thank you. Yes, you might try to set config.check_grad_for_nans to be True. On my datasets, I found few such cases. Here some my suggestions.

- Re-train it again and debug if it always raises such kind of errors.
- Set the depth loss in training code and confirm that if it comes from depth distillation loss or reconstruction loss. If it comes from depth distillation loss, try to reduce it and debug it.

Most of my code is from [RegNeRF](https://github.com/google-research/google-research/tree/master/regnerf).

**Comment by Wanggcong** [comment]
I am curious if your scenes are objects or larges-scale scenes. The sampling boxes are different for two cases. Please refer to this [link](https://github.com/Wanggcong/SparseNeRF/issues/7#issuecomment-1686215227) for the details.

**Comment by Wanggcong** [comment]
> Besides, I'm using my own dataset, not the example data you provided.

Hi, I write a rough tutorial in this [link](https://github.com/Wanggcong/SparseNeRF/blob/main/tutorial.md). Please let me know if it can solve your problem.

**Comment by hermosaaurora** [comment]
Can you explain why whether using NDC wrapping or not for such scenes can cause so different training results? If I don't use NDC, the loss can hardly decrease and the rendering results are insatisfactory (even worse than vanilla nerf)

In fact I implemented the depth rank loss in `nerf-pytorch` baseline and didn't see so much difference in training results if I do not use NDC wrapping.

**Comment by Wanggcong** [comment]
If you do not use NDC, what is the range of near and far?  You might print the reconstruction loss and depth loss to see the magnitudes of two losses. It seems that you might normalize the depth. You might try it.

**Comment by hermosaaurora** [comment]
near=0.2 and far=2 (that is precise because I have the ground truth z depth of the scene)
Besides, I'm a little confused about "It seems that you might normalize the depth".

**Comment by Wanggcong** [comment]
Basically, I am concerned about the balance between reconstruction loss and depth loss.

**Comment by hermosaaurora** [comment]
> Basically, I am concerned about the balance between reconstruction loss and depth loss.

hmm, I guess that's the crux. I will give it a try.

**Comment by Wanggcong** [comment]
Thank you. Please let me know if it works after your attempt.
